src/community/features-autopopulate/features-autopopulate-core/src/main/java/org/geoserver/autopopulate/AutopopulateTransactionCallback.java

```diff
@@ -0,0 +1,439 @@
+     *
+     * @see TransactionCallback#afterTransaction(TransactionRequest, TransactionResponse, boolean)
+     */
+    @Override
+    public void afterTransaction(
+            final TransactionRequest request, TransactionResponse result, boolean committed) {
+        // nothing to do
+    }
+
+    /**
+     * @return {@code 0}, we don't need any special treatment
+     * @see TransactionCallback#getPriority()
+     */
+    @Override
+    public int getPriority() {
+        return ExtensionPriority.LOWEST;
+    }
+
+    /**
+     * Collects "transactionCustomizer.properties" file that contains a set of names and CQL
+     * expressions, e.g.:
+     *
+     * <pre>
+     *  UTENTE=env('GSUSER')
+     *  AGGIORNAMENTO=now()
+     *  </pre>
+     *
+     * <p>To keep things simple, the expressions will just use environment variables, but not see
+     * the other values provided in the update/insert, and will not be differentiated by
+     * insert/update cases.
+     *
+     * @see TransactionListener#dataStoreChange(TransactionEvent)
+     */
+    @Override
+    public TransactionRequest beforeTransaction(TransactionRequest request) throws WFSException {
+        if (request.getVersion() == null) {
+            return request;
+        }
+
+        List<TransactionElement> newElements = new ArrayList<>();
+        for (TransactionElement element : request.getElements()) {
+            if (element instanceof Insert) {
+                List<SimpleFeature> newFeatures = new ArrayList<>();
+                for (Object of : affectedFeatures(element)) {
+                    if (of instanceof SimpleFeature) {
+                        try {
+                            log.info("Updating feature: " + of);
+                            SimpleFeature transformed = dataStoreChangeInternal((SimpleFeature) of);
+                            newFeatures.add(transformed);
+                        } catch (RuntimeException | IOException e) {
+                            // Do never make the transaction fail due to an
+                            // AutopopulateTransactionCallback error.
+                            // Yell on the logs though
+                            log.log(
+                                    Level.WARNING,
+                                    "Error pre computing the transaction's affected attributes",
+                                    e);
+                            newFeatures.add((SimpleFeature) of);
+                        }
+                    }
+                }
+                ((Insert) element).setFeatures(newFeatures);
+                newElements.add(element);
+            } else if (element instanceof Update) {
+                FeatureTypeInfo featureTypeInfo =
+                        getFeatureTypeInfo(new NameImpl(element.getTypeName()));
+                try {
+                    SimpleFeatureCollection features = getTransactionFeatures((Update) element);
```

Comment:
There are several casts of "element" to Update here, cast it just once in a local property instead.

Reply:
Done